server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java

```diff
@@ -0,0 +1,251 @@
+                        }
+
+                        @Override
+                        public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                            logger.debug("Removed repository cleanup task [{}] from cluster state", repositoryCleanupInProgress);
+                        }
+
+                        @Override
+                        public void onFailure(String source, Exception e) {
+                            logger.warn(
+                                "Failed to remove repository cleanup task [{}] from cluster state", repositoryCleanupInProgress);
+                        }
+                    });
+            }
+        });
+    }
+
+    private static ClusterState removeInProgressCleanup(final ClusterState currentState) {
+        RepositoryCleanupInProgress cleanupInProgress = currentState.custom(RepositoryCleanupInProgress.TYPE);
+        if (cleanupInProgress != null) {
+            boolean changed = false;
+            if (cleanupInProgress.cleanupInProgress() == false) {
+                cleanupInProgress = new RepositoryCleanupInProgress();
+                changed = true;
+            }
+            if (changed) {
+                return ClusterState.builder(currentState).putCustom(
+                    RepositoryCleanupInProgress.TYPE, cleanupInProgress).build();
+            }
+        }
+        return currentState;
+    }
+
+    @Override
+    protected CleanupRepositoryResponse read(StreamInput in) throws IOException {
+        return new CleanupRepositoryResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(Task task, CleanupRepositoryRequest request, ClusterState state,
+        ActionListener<CleanupRepositoryResponse> listener) {
+        if (state.nodes().getMinNodeVersion().onOrAfter(MIN_VERSION)) {
+            cleanupRepo(request.repository(), ActionListener.map(listener, CleanupRepositoryResponse::new));
+        } else {
+            throw new IllegalArgumentException("Repository cleanup is only supported from version [" + MIN_VERSION
+                + "] but the oldest node version in the cluster is [" + state.nodes().getMinNodeVersion() + ']');
+        }
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(CleanupRepositoryRequest request, ClusterState state) {
+        // Cluster is not affected but we look up repositories in metadata
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    /**
+     * Runs cleanup operations on the given repository.
+     * @param repositoryName Repository to clean up
+     * @param listener Listener for cleanup result
+     */
+    private void cleanupRepo(String repositoryName, ActionListener<RepositoryCleanupResult> listener) {
+        final Repository repository = repositoriesService.repository(repositoryName);
+        if (repository instanceof BlobStoreRepository == false) {
```

Comment:
We can check the read only repository upfront here

Reply:
You're right ... hmm ... the thing is, we do the check for `readOnly` for all the other operations in `BlobStoreRepository` as well. I can see how it's somewhat strange to go through the whole dance of updating the CS only to then fail because of the read only setting.
On the other hand, we do the check in `BlobStoreRepository` for create and delete and it seems safer to do that before doing any writing than relying on the caller of the cleanup method => maybe it's not worth the duplicate check to optimize the weird corner-case here? (mainly trying to keep the cleanup action code short)

Reply:
Since updating the cluster state can be a heavy task I really think we should fail early. But I'm surprised that we don't early check for read only when creating or deleting a snapshot.

So we could let this at it is in this PR and as a follow we could early check for read only repos in create/delete snapshot and repo clean up. WDYT?

Reply:
> So we could let this at it is in this PR and as a follow we could early check for read only repos in create/delete snapshot and repo clean up. WDYT?

Sounds good to me :) Def. makes things a little cleaner.